### PR TITLE
feat(core): add pure function combinators

### LINKED ISSUE #3481: pure functions to abstract and compose cloud resources
* **I'm submitting a ...**
  - [ ] :beetle: bug report
  - [x] :rocket: feature request
  - [ ] :books: construct library gap
  - [ ] :phone: security issue or vulnerability => Please see [policy](https://github.com/aws/aws-cdk/security/policy)
  - [ ] :question: support request => Please see note at the top of this template.

* **Please tell us about your environment:**

  - CDK CLI Version: 1.2.0 (build 6b763b7)
  - OS: OSX Mojave
  - Language: TypeScript

## Background

> Composition is the essence of programming...

The *composition* is a style of development to build a new things from small reusable elements. The composition is a key feature in functional programming. Functional code looks great only if functions clearly describe your problem. Usually lines of code per function is only a single metric that reflects quality of the code.

> If your functions have more than a few lines of code (a maximum of four to five lines per function is a good benchmark), you need to look more closely — chances are you have an opportunity to factor them into smaller, more tightly focused functions.

This is because functions describe the solution to your problem. If your code contains many lines then highly likely you are solving few problems without articulating them. A critical thinking is the process of software development. 

## Of the shelf composition with AWS CDK

A recent release of [AWS CDK](https://github.com/aws/aws-cdk) not only made an opportunity to deliver Infrastructure as a true Code but also gave a capability to make it from decomposable elements. Unfortunately, the development kit ignores a pure functional approach. The abstraction of cloud resources is exposed using class hierarchy.

> Constructs are the basic building blocks of AWS CDK apps. A construct represents a "cloud component" and encapsulates everything AWS CloudFormation needs to create the component.
> -- [AWS CDK](https://docs.aws.amazon.com/cdk/latest/guide/constructs.html)

Use this class to define code blocks that solves a single problem - just create a single cloud resource.

```typescript
import * as cdk from '@aws-cdk/core'

class MyResource extends cdk.Construct {
  constructor(scope: cdk.Construct, id: string) {
    super(scope, id)
    ...
  }
}

class MyStack extends cdk.Stack {
  constructor(app: cdk.App, id: string) {
    super(app, id)
    new MyResource(this, 'MyResource')
  }
}

const app = new cdk.App()
new MyStack(app, 'MyStack')
app.synth()
```

You'll be able to build reusable cloud component library and automate infrastructure patterns with real code. The usage of `cdk.Construct` only suffers from boilerplate code. 

## Pure functions

Let's try to define pure functions to solve embarrassingly obvious composition problem. We need to shift our perspective from category of classes to category of pure functions.

Let's abstract our Infrastructure as a Code in terms of pure functions. Functions that takes a parent cdk.Construct and creates a new element.

```typescript
type IaaC<T> = (parent: cdk.Construct) => T
```

The composition is about building efficiently parent-children relations due to nature of AWS CDK framework.

```typescript
type Compose = <T>(parent: cdk.Construct, children: IaaC<T>) => cdk.Construct
```

Finally, we are able to define `_` as compose function

```typescript
function _<T>(parent: cdk.Construct, fn: IaaC<T>): cdk.Construct {
  root instanceof cdk.App
    ? fn(new cdk.Stack(parent, fn.name))
    : fn(new cdk.Construct(parent, fn.name))
  return parent
}
```

The definition of our cloud resources becomes reflected in terms of pure functions

```typescript
function MyResource(scope: cdk.Construct): cdk.Construct {
   return ...
}

function MyStack(stack: cdk.Construct): cdk.Construct {
   return _(stack, MyResource)
}

const app = new cdk.App()
_(app, MyStack)
app.synth()
```

Now, the infrastructure is decomposable using small function. Each obviously correct and self explainable. This example eliminates boilerplate code of class constructions.

Here are some of the benefits:
* Infrastructure intent is clear and obvious
* It becomes testable and easier to reason about
* Easier to debug and maintain 

See the gist of [`pure.ts`](https://gist.github.com/fogfish/e97ef042db0afe011149873a56f79d93).

I'd be glad to hear your opinion on this subject.

### PR DESCRIPTION
**Pull request by fogfish**
closes #3481

The composition is a style of development to build a new things from small
reusable elements. The composition is a key feature in functional programming.
This change abstract our Infrastructure as a Code in terms of pure functions.

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

File: packages/@aws-cdk/core/test/test.pure.ts [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+//
+export = {
+  'join new IaaC block as a resource to stack'(test: Test) {
+    const stack = new Stack(undefined, 'MyStack');
+    pure.join(stack, IaaC);
+
+    test.deepEqual(toCloudFormation(stack),
+      {
+        Resources: {
+          IaaC59A6C06B: { Type: 'MyResourceType' }
```
Comment (fogfish) [inline packages/@aws-cdk/core/test/test.pure.ts]: What is a correct way to handle check of generated logical ids?

File: packages/@aws-cdk/core/lib/pure.ts [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+  }
+
+  /**
+   * lifts pure functional IaaC blocks to Stack node, then attaches then it to application.
+   * The node's logical name is either name of function or given name
+   * @param node
+   * @param fn
+   * @param name
+   */
+  export function add<T>(node: App, fn: IaaC<T>, name?: string): Construct {
```
Comment (fogfish) [inline packages/@aws-cdk/core/lib/pure.ts]: I've choose between `add` and `stack` name for the function.

**Comment by rix0rrr** [comment]
Thanks for the contribution!

I don't think this is appropriate to ship in the CDK core library, as (for example) it will only be available to TypeScript users -- this style of programming cannot be supported by [jsii](https://github.com/aws/jsii) -- and hence we cannot document or recommend it.

For now, we have chosen a different style of building CDK applications, on that translates to more common programming languages in a way that people are familiar with.

Having said that, I would totally recommend you build and publish this code as a separate library that works alongside the CDK to make it nicer for functional programmers. In sense, this is similar to what Sam has done with [Punchcard](https://github.com/sam-goodwin/punchcard). The beautiful thing about general purpose programming languages with package managers is that code doesn't have to live inside our code base to work together.

Best of luck!

**Comment by fogfish** [comment]
Thank you!  It's quite sad that you've provided this feedback as part of the #3481 discussion.

**Comment by rix0rrr** [comment]
Okay, maybe I spoke too hastily. From that thread, @eladb seems to think there is value in this.

Let's reopen the PR and assign it to him, he can have a look when he's back from holidays next week.

**Comment by rix0rrr** [comment]
I cannot reopen the PR, @fogfish can you do so?

**Comment by rix0rrr** [comment]
Oh, "the repository that submitted this pull request has been deleted"

**Comment by fogfish** [comment]
I've not seen any "hastily". It was very valid feedback. I am 100% agree about the fact that this pure functional aspect is not fitting very well to existing approach. Especially, the fact that it is only supported by TypeScript. 

The side library is a good approach to some extend but...  I still trying to design how to support the pure functional in best manner. E.g. I do not like the following construct because it requires kind of boilerplate of node construction `CfnResource(node, 'IaaC'...`

```typescript
function IaaC(node: Construct): Construct {
   return new CfnResource(node, 'IaaC', { type: 'MyResourceType' });
}
```
 
I really would like to see something like

```typescript
function IaaC(...): XxxTypeProps {
   return { type: 'MyResourceType' }
} 
```

and then when you join it materialises to correct node type XxxType

```typescript
pure.join<XxxType>(node, IaaC)
```

I think I've been rushing bit with PR. Let's discuss the design various proposals from the issue. I think I am missing yet a feedback from you side.

**Pull request closed without merging**